src/functions/dialog/index.ts

```diff
@@ -5,38 +5,80 @@
+	/**
+	 * Container to mount the dialog to
+	 * @default document.body
+	 */
+	container?: Element | string
 }
 
+export function spawnDialog(
+	dialog: Component,
+	props?: object,
+	onClose?: (...rest: unknown[]) => void,
+): Promise<unknown>
+
+export function spawnDialog(
+	dialog: Component,
+	props?: object,
+	options?: SpawnDialogOptions,
+	onClose?: (...rest: unknown[]) => void,
+): Promise<unknown>
+
 /**
- * Helper to spawn a Vue dialog without having to mount it from a component
+ * Spawn a single-use Vue dialog instance to get the result when it is closed
  *
- * @param dialog The dialog component to spawn
- * @param props Properties to pass to the dialog
- * @param props.container Optionally pass a query selector for the dialog container element
- * @param onClose Callback when the dialog is closed
+ * @param dialog - Dialog component to spawn
+ * @param props - Props to pass to the dialog instance
+ * @param optionsOrOnClose - Spawning options or a callback when the dialog is closed
+ * @param onClose - Callback when the dialog is closed
+ * @return Promise resolved with the `close` event payload
  */
 export function spawnDialog(
 	dialog: Component,
-	props?: DialogProps,
-	onClose: (...rest: unknown[]) => void = () => {},
-): void {
-	const el = document.createElement('div')
-	const container: HTMLElement = typeof props?.container === 'string'
-		? (document.querySelector(props.container) || document.body)
-		: document.body
-	container.appendChild(el)
-
-	const app = createApp(dialog, {
-		...props,
-		onClose: (...rest: unknown[]) => {
-			onClose(...rest.map(v => toRaw(v)))
-			app.unmount()
-			el.remove()
-		},
-	})
+	props: object = {},
+	optionsOrOnClose: SpawnDialogOptions | ((...rest: unknown[]) => void) = {},
+	onClose?: (...rest: unknown[]) => void,
+): Promise<unknown> {
+	if (typeof optionsOrOnClose === 'function') {
+		onClose = optionsOrOnClose
+		optionsOrOnClose = {}
+	}
 
-	app.mount(el)
+	let { container } = optionsOrOnClose
+
+	// For backwards compatibility try to use container from props
+	if ('container' in props && typeof props.container === 'string') {
+		container = props.container
+	}
+
+	// Resolve container to an Element or fallback to document.body
+	const resolvedContainer = (typeof container === 'string' ? document.querySelector(container) : container) || document.body
+
+	// Create root container element for the dialog
+	const element = resolvedContainer.appendChild(document.createElement('div'))
+
+	return new Promise((resolve, reject) => {
+		const app = createApp(dialog, {
+			...props,
+			// If dialog has no `container` prop passing a falsy value does nothing
+			// Otherwise it is expected that `null` disables teleport and mounts dialog in place like NcDialog/NcModal
+			container: null,
+			onClose(...rest: unknown[]) {
+				app.unmount()
+				element.remove()
+				const result = rest.map(v => toRaw(v))
+				onClose?.(...result)
+				resolve(result.length > 1 ? result : result[0])
```

Comment:
So you have to emit array or object your self?

Reply:
Yes. And in general, it's a common practice in Vue to always emit with one parameter in the payload. Then it's HTML-like and supports `$event` in template

Reply:
But here, if you emiteg many values - you got an array.

And if you emitted 1 value - you got it directly

Reply:
but you only resolve the first argument

Reply:
No, I only resolve the first argument when there are no arguments after the first

Reply:
I see :see_no_evil: